### PR TITLE
Allow using .ptr in @safe code when pointer is safe to dereference

### PR DESCRIPTION
**Pull request by ntrel**
Also parenthesize LHS expression in error message.
Allow for string literals.
Allow for array literals with length > 0.
~~Allow for compile-time length expression > 0.~~ - couldn't find any tests that needed this.

### CODE REVIEW

File: compiler/src/dmd/safe.d [hunk 1]
```diff
@@ -238,12 +238,20 @@ bool isSafeCast(Expression e, Type tfrom, Type tto)
  */
 bool checkUnsafeDotExp(Scope* sc, Expression e, Identifier id, int flag)
 {
-    if (!(flag & DotExpFlag.noDeref)) // this use is attempting a dereference
+    if (flag & DotExpFlag.noDeref) // this use is not attempting a dereference
+        return false;
+    if (id != Id.ptr)
+        return sc.setUnsafe(false, e.loc, "`(%s).%s` cannot be used in `@safe` code", e, id);
+
+    if (e.isStringExp()) // always null terminated
```
Comment (thewilsonator) [inline compiler/src/dmd/safe.d:246]: does this only get string literals? it looks like any string expressions which is not what we want.

Comment (RazvanN7) [reply compiler/src/dmd/safe.d:246]: String expressions are string literals.

**Comment by RazvanN7** [comment]
What's the motivation for this relaxation? Is it for passing D-style array/strings to C functions?

**Comment by WalterBright** [comment]
I have mixed feelings about this. While .ptr will indeed be safe if the expression is statically guaranteed to have at least one element, it adds a special case rule.

**Comment by Imperatorn** [comment]
> What's the motivation for this relaxation? Is it for passing D-style array/strings to C functions?

If that's the case I'd argue it's easier to just use a struct or library solution like cstring (it's just poc, but to communicate the point)

**Comment by RazvanN7** [comment]
Added the `72h no response -> close` label.

**Pull request closed without merging**